src/kmeans.jl

```diff
@@ -269,89 +280,82 @@ function update_centers!(
     end
 
     # sum ==> mean
-    @inbounds for j = 1:k
+    @inbounds for j ∈ 1:k
         if to_update[j]
-            for i = 1:d
-                centers[i, j] /= cweights[j]
+            cj = cweights[j]
+            for i ∈ 1:d
+                centers[i, j] /= cj
             end
         end
     end
 end
 
-#
-#  Update centers based on updated assignments
-#
-#  (specific to the case where samples are weighted)
-#
-function update_centers!(
-    X::AbstractMatrix{T},            # in: sample matrix (d x n)
-    weights::AbstractVector{<:Real}, # in: sample weights (n)
-    assignments::Vector{Int},        # in: assignments (n)
-    to_update::Vector{Bool},         # in: whether a center needs update (k)
-    centers::AbstractMatrix{T},      # out: updated centers (d x k)
-    cweights::Vector{Float64}        # out: updated cluster weights (k)
-    ) where T<:Real
+####
+#### update centers (weighted case)
+####
 
+function update_centers!(X::AbstractMatrix{TX}, weights::Vector{TW},
+                         assignments::Vector{Int}, to_update::Vector{Bool},
+                         centers::Matrix{TC}, cweights::Vector{TW}
+                         ) where {TX<:Real,TC<:AbstractFloat,TW<:Real}
     d, n = size(X)
     k = size(centers, 2)
 
     # initialize center weights
-    cweights[to_update] .= 0.0
+    cweights[to_update] .= zero(TW)
 
     # accumulate columns
-    @inbounds for j = 1:n
+    @inbounds for j ∈ 1:n
+        # skip samples with negative weights or assigned to a center
+        # that doesn't need to be updated
         wj = weights[j]
-        if wj > 0
-            cj = assignments[j]
-            1 <= cj <= k || error("assignment out of boundary.")
```

Comment:
Are you sure this never happens?

Reply:
yes, `assignments[j]` is set via the `argmin`  call which has for range something that is necessarily between `1` and `k` (otherwise one of the function calls throws an error earlier).

Reply:
It's worth to add a comment about that on `cj = assignments[j]` line: we're in `@inbounds` mode, so OOB indexing may crash Julia.

Reply:
ok to add a comment